draft-ietf-httpbis-connect-tcp.md

```diff
@@ -156,11 +157,71 @@ Clients SHOULD assume that all proxy resources generated by a single template sh
 
 ## Closing Connections
 
-In each HTTP version, any requirements related to closing connections in Classic HTTP CONNECT also apply to "connect-tcp", with the following modifications:
-
-* In HTTP/1.1, endpoints SHOULD close the connection in an error state to indicate receipt of a TCP connection error (e.g., a TCP RST or timeout).  Acceptable error states include sending an incomplete DATA capsule (as defined in {{Section 3.3 of !RFC9297}}), a TLS Error Alert ({{!RFC8446, Section 6.2}}), or a TCP RST (if TLS is not in use).  When a connection is terminated in an error state, the receiving endpoint SHOULD send a TCP RST if the underlying TCP implementation permits it.
-* In HTTP/2 and HTTP/3, senders MAY use an incomplete DATA capsule to indicate a TCP connection error, instead of (or in addition to) the signals defined for TCP connection errors in Classic HTTP CONNECT.  Recipients MUST recognize any incomplete capsule as a TCP connection error.
-* Intermediaries MUST propagate connection shutdown errors, including when translating between different HTTP versions.
+Connection termination is essentially symmetrical for proxies and their clients.  In this section, we use the term "endpoint" to describe an implementation of this specification in either role.
+
+When closing connections, endpoints are subject to the following requirements:
+
+* When an endpoint receives a valid TCP FIN, it MUST send a FINAL_DATA capsule.
+* When an endpoint receives a valid FINAL_DATA capsule, it MUST send a TCP FIN.
+* Each endpoint MUST close its send stream when the corresponding TCP connection reaches the TIME-WAIT or CLOSED state.
+  - If the connection closed gracefully, the endpoint MUST close the send stream gracefully.
+  - Otherwise, the endpoint SHOULD close the send stream abruptly, using a mechanism appropriate to the HTTP version:
+    - HTTP/3: RESET_STREAM with H3_CONNECT_ERROR
+    - HTTP/2: RST_STREAM with CONNECT_ERROR
+    - HTTP/1.1 over TLS: a TLS Error Alert
+    - HTTP/1.1 (insecure): TCP RST.
```

Comment:
Since we are giving advice on how to propagate TCP resets, do we want to RECOMMEND use of STOP_SENDING for H2/H3 too? I ask this because TCP resets are bidirectional.

Reply:
Nit: I believe STOP_SENDING is H3-only.

I think it's better not to send STOP_SENDING in this case.  Suppose we have [reliable reset](https://datatracker.ietf.org/doc/html/draft-ietf-quic-reliable-stream-reset), and some kind of POST_PROXY_SUMMARY capsule.  Sending STOP_SENDING would prevent delivery of POST_PROXY_SUMMARY in the reply, even though it might be sensible and useful.

Reply:
Yeah I think my point is that _if_ we are to give advise regarding how proxies should handle TCP resets, we ought to state that it is a bidirectional reset and that stream has to be tear down in both directions.

The mechanism can be different within each HTTP versions (or depending on the extensions in play).

Reply:
Ah, thank you very much for the explanation! I think I missed the part around the state machine. Looks good modulo editorial nits.